Add Link Time Optimization (LTO) option

ponchick opened the pull request:
Link Time Optimization (LTO) is very useful option especially on large libraries like mysensors. For example my LCD node sketch stats withoit LTO:
```
Sketch uses 30528 bytes (99%) of program storage space. Maximum is 30720 bytes.
Global variables use 1191 bytes (58%) of dynamic memory, leaving 857 bytes for local variables. Maximum is 2048 bytes.

```
with LTO enabled:
```
Sketch uses 27024 bytes (87%) of program storage space. Maximum is 30720 bytes.
Global variables use 1212 bytes (59%) of dynamic memory, leaving 836 bytes for local variables. Maximum is 2048 bytes.

```
Unfortunately SenseBender lacks that feature. That push request could fix that.

ponchick committed:
Add Link Time Optimization (LTO) option

fallberg commented:
This need to be supported in a ui less environment as well. Please make sure any additional arguments needed are added to .ci/arduino.groovy or toll-gate will not accept this pull request as it cannot be validated properly.

fallberg commented:
Also, if changes in the .ci files are needed in this repo, they will be a need for a corresponding change to the corresponding files in the MySensors repository as well.

ponchick committed:
Fix headless build

ponchick commented:
OK, I fixed uiless build. I changed .ci/arduino.groovy file as you suggested. Do I understand right that  I need to change only the function buildMySensorsMicro in the file .ci/arduino.groovy in the MySensors repository?

fallberg commented:
Yes, unfortunately. It is a hen-egg problem. When this pr is submitted, the jobs to validate the MySensors will pull updates from this repo (but use it's own groovy scripts) so it will fail for the same reason. It could be that (and I think it is) your change is such that you could already now file a PR to the MySensors repo to update the groovy beforehand without it failing so that the change can be submitted before this PR is submitted. That way, we can avoid breaking any builds in between.

Sorry for the inconvenience.
It was a intentional decision on my part to handle it like this because else you would have to update a third repo where the shared groovies would live, and then potentially all builds would break.

But it should be enough to only change the buildMySensorsMicro as it is the only one that deals with MySensortsMicro specific builds (which this PR is targeting).

ponchick commented:
Sorry, my English isn't good enough. My question was "Changes in function buildMySensorsMicro is the only changes I should make?". Ok, I'll create PR tomorrow morning.

fallberg commented:
I understood your question. Perhaps I wrote the answer a bit unclear. Yes, you should only need to change that function in the PR on the MySensors repository.

tekka007 commented:
LTO is enabled by default since Arduino 1.6.13 - I suggest updating _platform.txt_ to the most recent version (https://github.com/arduino/Arduino/blob/master/hardware/arduino/avr/platform.txt)

ponchick commented:
I prepared for pull request to the MySensors repository (development branch). If you'll decide to leave LTO optional I'm ready to create new PR.

tekka007 commented:
Fixed in #12
